at-http/src/main/java/io/engineblock/activitytypes/http/HttpAction.java

```diff
@@ -0,0 +1,189 @@
+
+import activityconfig.yaml.StmtDef;
+import activityconfig.yaml.StmtsBlock;
+import activityconfig.yaml.StmtsDoc;
+import activityconfig.yaml.StmtsDocList;
+import com.codahale.metrics.Timer;
+import io.engineblock.activityapi.core.ActivityDefObserver;
+import io.engineblock.activityapi.core.MultiPhaseAction;
+import io.engineblock.activityapi.core.SyncAction;
+import io.engineblock.activityimpl.ActivityDef;
+import io.virtdata.core.ContextualBindingsTemplate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.net.*;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+
+public class HttpAction implements SyncAction, MultiPhaseAction, ActivityDefObserver {
+    private final static Logger logger = LoggerFactory.getLogger(HttpAction.class);
+
+    private final ActivityDef activityDef;
+    private final HttpActivity httpActivity;
+    private final int slot;
+
+    // dynamic parameters
+    private int maxTries = 1;    // how many cycles a statement will be attempted for before giving up
+
+
+    private List<StmtDef> statements = new ArrayList<StmtDef>();
+    private boolean showstmnts;
+    private long nanoStartTime;
+    private ContextualBindingsTemplate contextualBindingsTemplate;
+    private StmtsDocList stmtDocList;
+
+
+    public HttpAction(ActivityDef activityDef, int slot, HttpActivity httpActivity) {
+        this.activityDef = activityDef;
+        this.httpActivity = httpActivity;
+        this.slot = slot;
+        onActivityDefUpdate(activityDef);
+    }
+
+    @Override
+    public void onActivityDefUpdate(ActivityDef activityDef) {
+        this.maxTries = httpActivity.getMaxTries();
+        this.showstmnts = httpActivity.getShowstmts();
+    }
+
+    @Override
+    public boolean incomplete() {
+        return false;
+    }
+
+    @Override
+    public int runPhase(long value) {
+        return runCycle(value);
+    }
+
+    @Override
+    public void init() {
+        onActivityDefUpdate(activityDef);
+
+        this.contextualBindingsTemplate = httpActivity.getContextualBindingsTemplate();
+        this.stmtDocList = httpActivity.getStmtDocList();
+        for (StmtsDoc doc : stmtDocList.getStmtDocs() ){
+            for (StmtsBlock stmtsBlock : doc.getBlocks()) {
+                List<StmtDef> stmtsFromBlock = stmtsBlock.getStmts();
+                statements.addAll(stmtsFromBlock);
+            }
+        }
+    }
+
+    @Override
+    public int runCycle(long cycleValue) {
+
+        String requestType = null;
+        String boundStatement = null;
+        InputStream result = null;
+
+        final Map<String, String> queryParamMap = new HashMap<String, String>();
+
+        int tries = 0;
+
+        try (Timer.Context bindTime = httpActivity.bindTimer.time()) {
+            if (statements.size() == 0){
+                logger.warn("No statements found.");
+            }
+            Iterator<StmtDef> it = statements.iterator();
```

Comment:
The selection of a bindable statement should be moved into initialization. If you look some uses of the Sequence planning, this will explain part of the pattern. The other part will be creating a statement template that only needs to have bind called on it to get the final statement form to execute. (Or string, or header, or request body, etc).
The binding logic and the sequencing logic should work together to limit the amount of work that has to happen within the cycle.